File: objc-project/Example/UnitTests/CoreFactoryTests.m
#import <Clappr/Clappr.h>

SPEC_BEGIN(CoreFactory)

describe(@"CoreFactory", ^{

    __block CLPPlayer *player;

    beforeEach(^{
        player = [CLPPlayer new];
    });

    describe(@"instantiation", ^{

        it(@"should raise an exception for its default initializer", ^{
            [[theBlock(^{
                [CLPCoreFactory new];
            }) should] raiseWithName:NSInternalInconsistencyException];
        });

        it(@"shoud have a designated initializer receiving the player", ^{

            CLPCoreFactory *factory = [[CLPCoreFactory alloc] initWithPlayer:player];

            [[factory.player should] equal:player];
        });
    });

    describe(@"core creation", ^{

        it(@"should be able to create a container using its create method", ^{
            CLPCoreFactory *factory = [[CLPCoreFactory alloc] initWithPlayer:player];
            CLPCore *core = [factory create];
            [[core shouldNot] beNil];
        });

        it(@"should be able to create a container with plugins from loader", ^{

            CLPCoreFactory *factory = [[CLPCoreFactory alloc] initWithPlayer:player];
            [[CLPLoader sharedLoader] stub:@selector(corePlugins) andReturn:@[[FakeUICorePlugin class]]];

            CLPCore *core = [factory create];

            BOOL containsPlugin = [core hasPlugin:[FakeUICorePlugin class]];
            [[theValue(containsPlugin) should] beTrue];
        });
    });

});

SPEC_END

File: Tests/CoreFactoryTests.swift
import Quick
import Nimble
import Clappr

class CoreFactoryTests: QuickSpec {
    
    override func spec() {
        describe("Core Factory") {
            context("Creation") {
                it("Should be able to create a core") {
                    let options = [kSourceUrl : "testUrl"]
                    let core = CoreFactory.create(options: options as Options)
                    
                    expect(core).toNot(beNil())
                    expect(core.container).toNot(beNil())
                }
                
                it("Should be able to create container with plugins") {
                    let loader = Loader(externalPlugins: [FakeUICorePlugin.self])

                    let core = CoreFactory.create(loader)
                    
                    expect(core.hasPlugin(FakeUICorePlugin)).to(beTrue())
                }

                it("Should add a core context to all plugins") {
                    let loader = Loader(externalPlugins: [FakeUICorePlugin.self])
                    let core = CoreFactory.create(loader)

                    expect(core.plugins).toNot(beEmpty())
                    for plugin in core.plugins {
                        expect(plugin.core) == core
                    }
                }
            }
        }
    }
    
    class FakeUICorePlugin: UICorePlugin {
        override var pluginName: String {
            return "FakeCorePLugin"
        }
    }
}
